core/src/main/java/io/grpc/internal/GzipInflatingBuffer.java

```diff
@@ -0,0 +1,514 @@
+          madeProgress = processHeaderMagic();
+          break;
+        case HEADER_COMPRESSION_METHOD:
+          madeProgress = processHeaderCompressionMethod();
+          break;
+        case HEADER_FLAG:
+          madeProgress = processHeaderFlag();
+          break;
+        case HEADER_EXTRA_LEN:
+          madeProgress = processHeaderExtraLen();
+          break;
+        case HEADER_EXTRA:
+          madeProgress = processHeaderExtra();
+          break;
+        case HEADER_NAME:
+          madeProgress = processHeaderName();
+          break;
+        case HEADER_COMMENT:
+          madeProgress = processHeaderComment();
+          break;
+        case HEADER_CRC:
+          madeProgress = processHeaderCrc();
+          break;
+        case INITIALIZE_INFLATER:
+          madeProgress = initializeInflater();
+          break;
+        case INFLATING:
+          madeProgress = inflate(bytesNeeded);
+          break;
+        case INFLATER_NEEDS_INPUT:
+          madeProgress = fill();
+          break;
+        case TRAILER:
+          madeProgress = processTrailer();
+          break;
+        default:
+          throw new AssertionError("Invalid state: " + state);
+      }
+    }
+    return inflaterOutputEnd;
+  }
+
+  private void writeInflatedBufToOutputBuffer(CompositeReadableBuffer bufferToWrite) {
+    bufferToWrite.addBuffer(ReadableBuffers.wrap(inflaterOutput, 0, inflaterOutputEnd));
+    inflaterOutputEnd = 0;
+    inflaterOutput = null;
+  }
+
+  private boolean inflate(int bytesToInflate) throws DataFormatException, ZipException {
+    checkState(inflater != null, "inflater is null");
+
+    try {
+      int inflaterTotalIn = inflater.getTotalIn();
+      int n = inflater.inflate(inflaterOutput, inflaterOutputEnd, bytesToInflate);
+      int bytesConsumedDelta = inflater.getTotalIn() - inflaterTotalIn;
+      bytesConsumed += bytesConsumedDelta;
+      inflaterInputStart += bytesConsumedDelta;
+      crc.update(inflaterOutput, inflaterOutputEnd, n);
+      inflaterOutputEnd += n;
+
+      if (inflater.finished()) {
+        // Save bytes written to check against the trailer ISIZE
+        expectedGzipTrailerIsize = (inflater.getBytesWritten() & 0xffffffffL);
+        if (gzipMetadataReader.readableBytes() <= GZIP_HEADER_MIN_SIZE + GZIP_TRAILER_SIZE) {
+          // We don't have enough bytes to begin inflating a concatenated gzip stream, drop context
+          inflater.end();
+          inflater = null;
+        }
+        state = State.TRAILER;
+        // Eagerly parse trailer, if possible, to detect CRC errors
+        return processTrailer();
+      } else if (inflater.needsInput()) {
+        state = State.INFLATER_NEEDS_INPUT;
+      }
+
+      return true;
+    } catch (DataFormatException e) {
+      // Wrap the exception so tests can check for a specific prefix
+      throw new DataFormatException("Inflater data format exception: " + e.getMessage());
+    }
+  }
+
+  private boolean initializeInflater() {
```

Comment:
Could you move this method to just before inflate(), so they are roughly in processing order/case order? (It doesn't need to be perfect, but this one is easy to fix)

Reply:
Done